### PR TITLE
Feature/iqmunicate 569

### PR DESCRIPTION
**Pull request by VitaliyGuru**

### CODE REVIEW

File: Q-municate/Classes/Core/QMCore/QMCore.m [hunk 1]
```diff
@@ -402,4 +403,26 @@ - (void)updateVocabulary {
...
+- (void)authServiceDidLogOut:(QMAuthService *)__unused authService {
+     NSLog(@"<PushNotificationManager> didLogout");
+    NSLog(@"<PushNotificationManager> enabled %@", self.currentProfile.pushNotificationsEnabled ? @"YES" : @"NO");
+    [[self.pushNotificationManager unregisterFromPushNotificationsAndUnsubscribe:YES] continueWithBlock:^id _Nullable(BFTask * _Nonnull __unused t) {
+        
+        return nil;
+    }];
+}
+
+//- (void)authService:(QMAuthService *)__unused authService
```
Comment (Pro100Andrey) [inline Q-municate/Classes/Core/QMCore/QMCore.m]: ???

File: Q-municate/Classes/Core/QMPushNotificationManager/QMPushNotificationManager.m [hunk 2]
```diff
@@ -12,51 +12,193 @@
...
 
 @end
 
 @implementation QMPushNotificationManager
 
 @dynamic serviceManager;
 
+
+- (BFTask *)unregisterFromPushNotificationsAndUnsubscribe:(BOOL)shouldUnsubscribe {
+
```
Comment (Pro100Andrey) [inline Q-municate/Classes/Core/QMPushNotificationManager/QMPushNotificationManager.m]: should be:

 [[UIApplication sharedApplication] unregisterForRemoteNotifications];

[[self unSubscribeFromPushNotifications] continueWithBlock:^id _Nullable(BFTask * _Nonnull t {
             if (t.isCancelled) {
                 [source trySetCancelled];
             }
             else if (t.error) {
                 [source setError:t.error];
         }];

File: Q-municate/Scenes/Auth/Login screen/QMLogInViewController.m [hunk 3]
```diff
@@ -74,7 +74,7 @@ - (IBAction)done:(id)__unused sender {
                 QMCore.instance.currentProfile.accountType = QMAccountTypeEmail;
                 [QMCore.instance.currentProfile synchronizeWithUserData:task.result];
                 
-                [QMCore.instance.pushNotificationManager registerForPushNotifications];
+              //  [QMCore.instance.pushNotificationManager registerForPushNotifications];
```
Comment (Pro100Andrey) [inline Q-municate/Scenes/Auth/Login screen/QMLogInViewController.m]: remove  unused code

File: Q-municate/Scenes/Auth/Welcome screen/QMWelcomeScreenViewController.m [hunk 4]
```diff
@@ -126,7 +126,7 @@ - (void)chainFacebookConnect {
                 }];
             }
             
-             [QMCore.instance.pushNotificationManager registerForPushNotifications];
+           //  [QMCore.instance.pushNotificationManager registerForPushNotifications];
```
Comment (Pro100Andrey) [inline Q-municate/Scenes/Auth/Welcome screen/QMWelcomeScreenViewController.m]: remove unused code

File: Q-municate/Scenes/Settings/QMSettingsViewController/QMSettingsViewController.m [hunk 5]
```diff
@@ -153,20 +153,28 @@ - (IBAction)pushNotificationSwitchPressed:(UISwitch *)sender {
     [(QMNavigationController *)self.navigationController showNotificationWithType:QMNotificationPanelTypeLoading message:NSLocalizedString(@"QM_STR_LOADING", nil) duration:0];
     
     __weak QMNavigationController *navigationController = (QMNavigationController *)self.navigationController;
+    
     BFContinuationBlock completionBlock = ^id _Nullable(BFTask * _Nonnull __unused task) {
+        if (task.faulted) {
+            [navigationController showNotificationWithType:QMNotificationPanelTypeFailed message:task.result duration:3];
+        }
+        else {
+            [QMCore instance].currentProfile.pushNotificationsEnabled = ![QMCore instance].currentProfile.pushNotificationsEnabled;
```
Comment (Pro100Andrey) [inline Q-municate/Scenes/Settings/QMSettingsViewController/QMSettingsViewController.m]: ??

**Comment by Pro100Andrey** [comment]
update clearProfile method, 
- (BOOL)clearProfile {
    
    __block BOOL success = NO;
    
    [self keychainQuery:^(SSKeychainQuery *query) {
        
        NSError *error = nil;
        success = [query deleteItem:&error];
    }];
    
    self.userData = nil;
    self.accountType = QMAccountTypeNone;
    self.pushNotificationsEnabled = YES;
    self.userAgreementAccepted = NO;
    self.lastDialogsFetchingDate = nil;
    self.lastUserFetchDate = nil;
    
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    [defaults setBool:NO forKey:kQMAppExists];
    [defaults synchronize];
    
    return success;
}

**Comment by Pro100Andrey** [comment]
Update  synchronize in QMProfile
- (BOOL)synchronize {
     NSParameterAssert(self.userData);
     
     __block BOOL success = NO;
     __block NSError *error = nil;
     
     @weakify(self);
     [self keychainQuery:^(SSKeychainQuery *query) {
         @strongify(self);
         NSCParameterAssert(self.userData.password);
         query.passwordObject = self;
         success = [query save:&error];
     }];
     
     if (success) {
         
         NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
         [defaults setBool:YES forKey:kQMAppExists];
-        
+        [defaults synchronize];
         // updating user in users cache
         [QMCore.instance.usersService.usersMemoryStorage addUser:self.userData];
         [[QMUsersCache instance] insertOrUpdateUser:self.userData];
     }
     else {
         NSLog(@"QMProfile error %@", error);
     }
     
     return success;
 }

File: Q-municate/Classes/Core/QMPushNotificationManager/QMPushNotificationManager.m (hunk 2 above)
Comment (VitaliyGuru) [inline Q-municate/Classes/Core/QMPushNotificationManager/QMPushNotificationManager.m]: when should it call unsubscribe later in case of error?

**Pull request merged**